Topic/v5.x/oshmem/communicator mem footprint fix

sb17v opened the pull request.

skosukhin committed:
Update Libtool patch for NAG.

Signed-off-by: Sergey Kosukhin <sergey.kosukhin@mpimet.mpg.de>

ggouaillardet committed:
configury: add missing documentation to OPAL_CHECK_PACKAGE() parameter

No code change

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

skosukhin committed:
Enable convenience libraries for NAG.

Signed-off-by: Sergey Kosukhin <sergey.kosukhin@mpimet.mpg.de>

bureddy committed:
CUDA: Enhance MPIX_Query_cuda_support() with runtime support check

Signed-off-by: Devendar Bureddy <devendar@mellanox.com>

abouteiller committed:
Coverity CID 1472864 is invalid; add an assert to make this more
explicit

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity 1472863: useless compare with 0 for unsigned

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity CID 1472866: missing return in mca_base_var_load_extra_files

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity CID 1472867 and 1472870: changing recast method so as to not
confuse coverity as much

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity CID 1472861: dereferencing NULL groups was indeed incorrect

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity CID 1472860: ci is not NULL, add an assert to help coverity see
it

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity CID 1472869: add an assert to let coverity know that its ok
av->new_dead_array is dereferenced only if nb_new_dead>0

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
coverity 1472872: check for NULL return from strdup

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

bwbarrett committed:
build: remove unused configuration from mpisync

The mpisync application does not use any of the CPPFLAGS that were
being explicitly added in the Makefile.am.  This looks like copy
and paste extras from ompi_info.

Signed-off-by: Brian Barrett <bbarrett@amazon.com>

bwbarrett committed:
Do not use full path for wrapper C compiler

Fix an assymetry in the wrapper compilers, where the C compiler
was always specified as a full path (regardless of the value
of $CC), while C++ and Fortran compilers were always invoked
based on the value of $CXX and $FC.  In the case where the
user directly sets $CC before invoking configure, this will
result in the C compiler invoked by the wrapper compiler
exactly matching $CC.

Signed-off-by: Brian Barrett <bbarrett@amazon.com>

bwbarrett committed:
build: Remove opalc++ wrapper compiler

We no longer initialize the C++ compiler if we are not building
the MPI or SHMEM layers and no longer require a C++ compiler.
Rather than handling all the special cases to build the opalc++
wrapper compiler that no one uses, kill it off.

Signed-off-by: Brian Barrett <bbarrett@amazon.com>

devreal committed:
Have opal_progress() return the number of events it encountered

This information can used by callers to decide how to proceed
based on whether operations have progressed.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
MPI_Test: only recheck if opal_progress signaled events

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

devreal committed:
MPI_Testall: trigger progress if encountering an incomplete request and short-circuit

While iterating over the list of requests, the first incomplete request
should trigger the progress engine. Subsequent incomplete encounters
of incomplete requests allow us to short-circuit the iteration
and return early.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

rhc54 committed:
Create an mpirun dedicated executable

If we install the embedded versions or PRRTE and PMIx, then
OPAL_PREFIX needs to adjust the installed location for PMIx
as well as for PRRTE when launching via mpirun.

In mpirun: Detect OPAL_PREFIX and set the corresponding PRRTE and PMIx
prefixes. Take advantage of the simplified proxy detection
support in PRRTE. Deal with absolute path names

In MPI_Init: Detect OPAL_PREFIX and set the corresponding PMIx
prefix

Signed-off-by: Ralph Castain <rhc@pmix.org>

abouteiller committed:
fix hugepage allocator assert in finalize

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
Coverity 1472868: members cannot be NULL at this point.

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

abouteiller committed:
Coverity 1472875 (and friends): use a direct rather than indirect test
for accessing value->bytes, so that we don't obfuscate from coverity

Signed-off-by: Aurelien Bouteiller <bouteill@icl.utk.edu>

sb17v commented on oshmem/shmem/c/shmem_ptr.c:
```diff
@@ -46,8 +46,8 @@ void *shmem_ptr(const void *dst_addr, int pe)
     /* The memory must be on the local node */
-    proc = oshmem_proc_group_find(oshmem_group_all, pe);
-    if (!OPAL_PROC_ON_LOCAL_NODE(proc->super.proc_flags)) {
+    proc_flag = oshmem_proc_group_find_v2(oshmem_group_all, pe);
+    if (!OPAL_PROC_ON_LOCAL_NODE(proc_flag)) {
```
```suggestion
    proc_flag = oshmem_group_all->proc_locality.bitmap[pe];
    if (!OPAL_PROC_ON_LOCAL_NODE(proc_flag)) {
```

sb17v commented on oshmem/proc/proc.h:
```diff
@@ -301,18 +340,14 @@ static inline int oshmem_my_proc_id(void)
-    proc = oshmem_proc_group_find(oshmem_group_all, pe);
-
-    return (int) OSHMEM_PROC_DATA(proc)->transport_ids[0];
+    opal_hwloc_locality_t proc_flag;
```
```suggestion
```

sb17v commented on oshmem/proc/proc.c:
```diff
@@ -252,8 +304,16 @@ oshmem_proc_group_destroy_internal(oshmem_group_t* group, int scoll_unselect)
+    }
+
+    if (group->proc_jobids) {
+        free(group->proc_jobids);
```
Release the proc_locality here

sb17v commented on oshmem/proc/proc.c:
```diff
@@ -181,27 +228,32 @@ oshmem_group_t* oshmem_proc_group_create(int pe_start, int pe_stride, int pe_siz
                 && ((pe_stride == 0)
                     || (((cur_pe - pe_start) % pe_stride) == 0))) {
-            proc_array[count_pe++] = proc;
+            
+            // proc_array[count_pe++] = proc;
+            // struct group {
+            //     ....
+            //     opal_vpid_t *proc_vpids; //[group_size];
+            //     opal_bitmap_t proc_locality[group_size/8];
+            //     ...
+            // } 
+
```
```suggestion
```

sb17v commented on oshmem/proc/proc.c:
```diff
@@ -27,11 +27,12 @@
 
 static opal_mutex_t oshmem_proc_lock;
-
+static opal_bitmap_t _oshmem_local_vpids;
```
```suggestion
```

sb17v commented on oshmem/proc/proc.c:
```diff
@@ -80,12 +84,51 @@ int oshmem_proc_group_init(void)
+    }
+    rc = opal_bitmap_init(&oshmem_group_self->proc_locality, 1);
+    if (OMPI_SUCCESS != rc) {
+        goto err2;
```
```suggestion
        goto err3;
```

ggouaillardet committed:
configury: cleanup the configury logic for ime, lustre and pvfs2

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet committed:
fs/ime: fix an abstraction violation

close IME when the fs/ime component is closed,
and not in io/ompio since that would be an abstraction violation.

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

devreal committed:
MPI_Testsome: don't check request for failed state if found complete

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

sb17v committed:
oshmem: Reduce the memory footprint for communicator creation

oshmem uses ompi_proc structure during group ceation. However,
only vpids and locality information is used inside oshmem internally.
Removing the proc_array from oshmem_group and replacing it with a
array to hold the vpids. A static array is used to keep track of
locality information which we collect from PMIx.

Example: The original ompi_proc_t structure in proc_array was about 112 bytes,
while each entry in vpid array will be 4 byte. We use a bitmap
to track locality information which is around 1/8 bytes. So,
we reduces the memory usage from 112 to ~4.13 per proc in each pe.
With 100 nodes and 40 PPN, each proc used to utilize (112*40) = 4480 bytes.
Now, after this PR, it can come down to (4.13 * 40) = 165 bytes.
So, the total memory utilization per node (w/ 40 ppn) in this codeflow
will come down from 179,200 bytes to 6,600 bytes.

Co-authored-by: Artem Y. Polyakov <artemp@nvidia.com>
Signed-off-by: Subhadeep Bhattacharya <subhadeepb@nvidia.com>
Signed-off-by: Artem Polyakov <artemp@nvidia.com>

artpol84 commented on oshmem/proc/proc.c:
```diff
@@ -34,7 +34,34 @@ static opal_mutex_t oshmem_proc_lock;
     OBJ_CONSTRUCT(&oshmem_proc_lock, opal_mutex_t);
+    OBJ_CONSTRUCT(&_oshmem_local_vpids, opal_bitmap_t);
+
+    ret = opal_bitmap_init(&_oshmem_local_vpids, ompi_comm_size(oshmem_comm_world));
+    if (OMPI_SUCCESS != ret) {
+        return ret;
+    }
+    /* Add all local peers first */
+    wildcard_rank.jobid = OMPI_PROC_MY_NAME->jobid;
+    wildcard_rank.vpid = OMPI_NAME_WILDCARD->vpid;
+    /* retrieve the local peers */
+    OPAL_MODEX_RECV_VALUE(ret, PMIX_LOCAL_PEERS,
+                          &wildcard_rank, &val, PMIX_STRING);
+
+    if (OPAL_SUCCESS == ret && NULL != val) {
+        char **peers = opal_argv_split(val, ',');
+        int i;
+        free(val);
+        for (i=0; NULL != peers[i]; i++) {
+            ompi_vpid_t local_rank = strtoul(peers[i], NULL, 10);
+            opal_bitmap_set_bit(&_oshmem_local_vpids, local_rank);
+        }
+        opal_argv_free(peers);
+    }
```
better make it a separate function

artpol84 commented on oshmem/proc/proc.h:
```diff
@@ -149,14 +150,17 @@ static inline ompi_proc_t *oshmem_proc_find(int pe)
 
+static inline int oshmem_proc_pe_vpid(oshmem_group_t *group, int pe)
+{
+    return ((pe < group->proc_count) ? (group->proc_vpids[pe]): -1);
```
```suggestion
    if (OPAL_LIKELY(pe < group->proc_count)) {
        return (group->proc_vpids[pe]);
    } else {
        return 1;
    }
```

artpol84 commented on oshmem/mca/memheap/base/memheap_base_mkey.c:
```diff
@@ -168,7 +165,7 @@ static void unpack_remote_mkeys(shmem_ctx_t ctx, pmix_data_buffer_t *msg, int re
             cnt = 1;
             PMIx_Data_unpack(NULL, msg, &memheap_oob.mkeys[tr_id].u.key, &cnt, PMIX_UINT64);
-            if (OPAL_PROC_ON_LOCAL_NODE(proc->super.proc_flags)) {
+            if (OSHMEM_PROC_ON_LOCAL_NODE(remote_pe)) {
```
do we have `remote_pe` = global PE here?

artpol84 commented on oshmem/proc/proc.h:
```diff
@@ -149,14 +151,50 @@ static inline ompi_proc_t *oshmem_proc_find(int pe)
-#define OSHMEM_PROC_VPID(PROC)     (((ompi_process_name_t*)&((PROC)->super.proc_name))->vpid)
-
+#define OSHMEM_PROC_ON_LOCAL_NODE(pe) (opal_bitmap_is_set_bit(&_oshmem_local_vpids, pe))
+static inline int oshmem_proc_init_set_local_vpids()
```
Why do we need this functions as inline?
It is not on the data path and doesn't need to be fast.

sb17v replied:
```suggestion
int oshmem_proc_init_set_local_vpids();
```

artpol84 commented on oshmem/proc/proc.h:
```diff
@@ -55,14 +57,14 @@ typedef struct oshmem_proc_data_t oshmem_proc_data_t;
  *
  * Set of processes used in collective operations.
  */
+static opal_bitmap_t _oshmem_local_vpids;       /* Track the vpids in local node */
```
This array should not be in the header file. Static var doesn't make sense in the header.
The static bitmap is declared in the *.c file and then functions to access it are declared here.

sb17v replied:
```suggestion
```

sb17v commented on oshmem/proc/proc.c:
```diff
@@ -34,7 +34,14 @@ static opal_mutex_t oshmem_proc_lock;
 
```
```suggestion
static opal_bitmap_t _oshmem_local_vpids;       /* Track the vpids in local node */
int oshmem_proc_init_set_local_vpids()
{
    opal_process_name_t wildcard_rank;
    int ret = OMPI_SUCCESS;
    char *val = NULL;
    
    ret = opal_bitmap_init(&_oshmem_local_vpids, ompi_comm_size(oshmem_comm_world));
    if (OSHMEM_SUCCESS != ret) {
        return ret;
    }
    /* Add all local peers first */
    wildcard_rank.jobid = OMPI_PROC_MY_NAME->jobid;
    wildcard_rank.vpid = OMPI_NAME_WILDCARD->vpid;
    /* retrieve the local peers */
    OPAL_MODEX_RECV_VALUE(ret, PMIX_LOCAL_PEERS,
                          &wildcard_rank, &val, PMIX_STRING);

    if (OPAL_SUCCESS == ret && NULL != val) {
        char **peers = opal_argv_split(val, ',');
        int i;
        free(val);
        for (i=0; NULL != peers[i]; i++) {
            ompi_vpid_t local_rank = strtoul(peers[i], NULL, 10);
            opal_bitmap_set_bit(&_oshmem_local_vpids, local_rank);
        }
        opal_argv_free(peers);
    }
    return OSHMEM_SUCCESS;
}
```

sb17v commented on oshmem/proc/proc.h:
```diff
@@ -149,14 +151,50 @@ static inline ompi_proc_t *oshmem_proc_find(int pe)
-
+#define OSHMEM_PROC_ON_LOCAL_NODE(pe) (opal_bitmap_is_set_bit(&_oshmem_local_vpids, pe))
+static inline int oshmem_proc_init_set_local_vpids()
+{
+    opal_process_name_t wildcard_rank;
+    int ret = OMPI_SUCCESS;
+    char *val = NULL;
+    
+    ret = opal_bitmap_init(&_oshmem_local_vpids, ompi_comm_size(oshmem_comm_world));
+    if (OSHMEM_SUCCESS != ret) {
+        return ret;
+    }
+    /* Add all local peers first */
+    wildcard_rank.jobid = OMPI_PROC_MY_NAME->jobid;
+    wildcard_rank.vpid = OMPI_NAME_WILDCARD->vpid;
+    /* retrieve the local peers */
+    OPAL_MODEX_RECV_VALUE(ret, PMIX_LOCAL_PEERS,
+                          &wildcard_rank, &val, PMIX_STRING);
+
+    if (OPAL_SUCCESS == ret && NULL != val) {
+        char **peers = opal_argv_split(val, ',');
+        int i;
+        free(val);
+        for (i=0; NULL != peers[i]; i++) {
+            ompi_vpid_t local_rank = strtoul(peers[i], NULL, 10);
+            opal_bitmap_set_bit(&_oshmem_local_vpids, local_rank);
+        }
+        opal_argv_free(peers);
+    }
+    return OSHMEM_SUCCESS;
+}
```
```suggestion
```

rhc54 committed:
Update PMIx and PRRTE pointers

Pickup transfer of output formatting from PRRTE to PMIx

Signed-off-by: Ralph Castain <rhc@pmix.org>

gpaulsen committed:
Merge pull request #9095 from rhc54/topic/up

Update PMIx and PRRTE pointers

hppritcha committed:
LANL-CI: run osu tests which exercise CUDA

Open MPI interoperability

Signed-off-by: Howard Pritchard <howardp@lanl.gov>

sb17v committed:
oshmem: Remove dead codes during oshmem communicator creation

oshmem uses oshmem_group_t structure during communicator creation.
However, it consists a peer_list of type opal_list_t. However, the
peer_list field is not used anywhere inside oshmem. Removing the
peer_list field from the oshmem_group_t structure to save additional
memory consumption.

Co-authored-by: Artem Y. Polyakov <artemp@nvidia.com>
Signed-off-by: Subhadeep Bhattacharya <subhadeepb@nvidia.com>
Signed-off-by: Artem Polyakov <artemp@nvidia.com>

Commit:
ofi common: add "usnic" to the exclude list

We never want to use usnic devices with the OFI MTL or OFI BTL (we
have the usnic BTL, instead).  So add "usnic" to the default common
OFI exclude list so that neither "usnic" nor "usnic;ofi_rxd" will be
used by the OFI MTL.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

Commit:
ofi: fix compile warnings

Properly prefix global symbols, thereby also fixing some "parameter
shadows global variable name" warnings.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

Commit:
usnic: fix some compiler warnings

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

jsquyres committed:
Merge pull request #9101 from jsquyres/pr/ofi-fixes

OFI and usNIC fixes

hppritcha committed:
Merge pull request #9100 from hppritcha/topic/add_ibm_cuda_xl_p9_plus_tests

LANL-CI: run osu tests which exercise CUDA

awlauria committed:
Merge pull request #8486 from abouteiller/coverity/ulfm

ULFM followup: coverity reports

hppritcha committed:
Merge pull request #9069 from abouteiller/fix/finalize/hugeallocator-mpool

fix hugepage allocator assert in finalize

awlauria committed:
Remove .travis.yml as it's no longer used with CI.

Can always bring this back if that fact ever changes.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

awlauria committed:
Merge pull request #7970 from bureddy/cuda-query

CUDA: Enhance MPIX_Query_cuda_support() with runtime support check

awlauria committed:
Merge pull request #9058 from sb17v/topic/v5.x/oshmem/mem_footprint_fix_v2

oshmem: Reduce the service memory footprint

samuelkgutierrez committed:
OFI: Fix misleading component selection debug output.

* Make explicit that a provider's NIC is displayed.
* Normalize naming convention in debug output.

Signed-off-by: Samuel K. Gutierrez <samuel@lanl.gov>

samuelkgutierrez committed:
Rename provider:nic to provider:domain.

Signed-off-by: Samuel K. Gutierrez <samuel@lanl.gov>

awlauria committed:
Merge pull request #9110 from awlauria/remove_travis_yml

Remove .travis.yml as it's no longer used with CI.

samuelkgutierrez committed:
Normalize mtl:ofi opal_output_verbose() sinks.

Signed-off-by: Samuel K. Gutierrez <samuel@lanl.gov>

samuelkgutierrez committed:
edit .mailmap file to resolve Jenkins warnings.

Signed-off-by: Samuel K. Gutierrez <samuel@lanl.gov>

rhc54 committed:
Merge pull request #9108 from rhc54/topic/mpirun

Create an mpirun dedicated executable

gpaulsen committed:
Merge pull request #6389 from ggouaillardet/topic/configury_misc_fixes

configury: misc fixes

awlauria committed:
Merge pull request #8904 from devreal/mpi-test-recheck

Improvements to MPI_Test and friends

gpaulsen committed:
Merge pull request #8793 from bwbarrett/bugfix/mpicc-should-use-the-same-compiler-as-ompi

Wrapper compiler cleanups

gpaulsen committed:
Merge pull request #9111 from samuelkgutierrez/ofi-fix-debug

OFI MTL: Fix component selection debug output

gpaulsen committed:
Merge pull request #6378 from skosukhin/fix_nag_pthread

Fix inherited_linker_flags processing when building with NAG compiler

rhc54 committed:
Get the correct directory when using installdirs

Need to use the bindir entry to find the "prterun" executable.
Add an output line indicating that execve failed and why.
Return non-zero status on exit.

Signed-off-by: Ralph Castain <rhc@pmix.org>

rhc54 committed:
Merge pull request #9116 from rhc54/topic/exec

Get the correct directory when using installdirs

awlauria committed:
Add a mfence to opal_atomic_mb() for x86.

x86 requires a mfence for it to be a full memory barrier;
a compiler MB is not sufficient.

Co-authored-by: George Katevenis <george_kate@hotmail.com>

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

bwbarrett committed:
dist: Sync NEWS from release branches

Pull in release information from both the v4.0.x and v4.1.x relaase
branches.

Signed-off-by: Brian Barrett <bbarrett@amazon.com>

awlauria committed:
Fix --prefix option in mpirun.

Tack "/bin" back on the end of the --prefix line to preserve old
behavior.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

bwbarrett committed:
Merge pull request #9134 from bwbarrett/dist/master-NEWS-sync

dist: Sync NEWS from release branches

awlauria committed:
Merge pull request #9130 from awlauria/fix_x86_mb

Add a mfence to opal_atomic_mb() for x86.

bwbarrett committed:
Merge pull request #9137 from bwbarrett/feature/clean-up-mpisync-makefile

build: remove unused configuration from mpisync

awlauria committed:
Merge pull request #9136 from awlauria/fix_prefix_master

Fix --prefix option in mpirun.

Commit:
coll/sm: fix Powerpoint design file

Somewhere in its history, the memory-layout.ppt file got corrupted
here in the Git repo.  Restore an old copy of this file, and upgrade
it to the modern .pptx format.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

ggouaillardet committed:
romio314: do not link with CUDA

ROMIO internally relies on MPL, that links to CUDA if detected
at configure time.
Not only this is not required by ROMIO (at least for now), but
this is not a fit for Open MPI.
This commit forces ROMIO *not* to detect CUDA at configure time
by using setting the hidden/undocumented
ac_cv_lib_cuda_cuMemGetAddressRange=no
environment variable before invoking ROMIO's configure.

Refs. open-mpi/ompi#8884

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet committed:
Merge pull request #9144 from ggouaillardet/topic/cudaless_romio

romio314: do not link with CUDA

hjelmn committed:
test: Update Makefile.am files to not define distclean target

This commit fixes warnings issued by automake about overriding the builtin distclean target.
The fix is to use distclean-local instead.

Signed-off-by: Nathan Hjelm <hjelmn@google.com>

hjelmn committed:
Merge pull request #9140 from jsquyres/pr/fix-memory-layout-ppt-file

coll/sm: fix Powerpoint design file

hjelmn committed:
Merge pull request #9147 from hjelmn/automake_cleanup

test: Update Makefile.am files to not define distclean target
